### DISCUSSION TITLE
Communicate mode switch to external apps

### DISCUSSION THREAD

**Discussion by eugenesvk** [helix-editor/helix#6081] [Q&A]
I'd like to communicate Helix's mode switch to external apps, e.g., to the terminal to disable some keybinds, and to a key rebinding app like Karabiner Elements for the same reason

I can do one piece of it: run a shell command to set the Karabiner variable (e.g., `🧬i` for insert mode)
But then with the terminal in a regular shell session I could run a special escape sequence to set a user variable (`printf '\033]1337;SetUserVar=🧬i=MQ==\007'`, which generates an event I can then use to change keybinds), but then it does nothing when run inside Helix's run-shell-command `:sh` (maybe it's because stdout is redirected to Helix's panel and doesn't reach the terminal?? I tried escaping the `\`, but that didn't help)

Also, this might be an unreliable and tedious rebinding exercise as I'd need to remap every key that switches modes into a `['my long commands that send signal to external apps', 'original command']` sequence, but even then, e.g., how would I track the end of the command mode?

So I have a couple of questions:
1. Is it possible to somehow send the OSC 1337-wrapped command so that the terminal can pick it up?
2. Is it possible to "hook" into any and all mode switches and ask Helix to run a couple of commands that would communicate this switch to other apps
3. Is there a better way for this communication to happen?

**Comment by kirawi** [comment]
This would've been likely implemented as part of a Kakoune-like hook system, but AFAIK the plugin system takes precedence over that.
